src/Umbraco.Core/Persistence/Repositories/Implement/ContentRepositoryBase.cs

```diff
@@ -630,6 +630,13 @@ private static void AppendNodeToFix(IDictionary<int, List<NodeDto>> nodesToRebui
             }
             if (versions.Count == 0) return new Dictionary<int, PropertyCollection>();
 
+            // TODO: This is a bugger of a query and I believe is the main issue with regards to SQL performance drain when querying content
+            // which is done when rebuilding caches/indexes/etc... in bulk. We are using an "IN" query on umbracoPropertyData.VersionId
+            // which then performs a Clustered Index Scan on PK_umbracoPropertyData which means it iterates the entire table which can be enormous!
+            // especially if there are both a lot of content but worse if there is a lot of versions of that content.
+            // So is it possible to return this property data without doing an index scan on PK_umbracoPropertyData and without iterating every row
+            // in the table?
+
             // get all PropertyDataDto for all definitions / versions
             var allPropertyDataDtos = Database.FetchByGroups<PropertyDataDto, int>(versions, 2000, batch =>
```

Comment:
I don't think this is called during startup, just from the rebuild database cache. 
Changing the batch size to 500 results in a index seek instead of an index scan. This call often causes timeouts. 

I think this method could be reordered. It may be better to query the database first for the PropertyTypeDto's rather than getting the ids from the propertyDataDtos. This method is also really memory expensive as it loads all published and new versions of all content into memory at the same time, it may be better to page through them, adding them to the dictionary to smooth out the load on the database server.

Reply:
> I don't think this is called during startup, just from the rebuild database cache.

Yep this is called whenever IContent or collections of it is queried. The grouped results to 500 using an index seek is interesting though. Do you know if there's a certain limit to when it decides to start scanning instead of seeking?

Changing this to not load everything into memory could only be done by refactoring a lot of how this method gets used in the first place. For queries where tons are returned - like when rebuilding the db cache or when displaying content in list views it could be interesting to see how this could be improved. 

For now however do you think it's worth just changing the 2000 -> 500 ?

Reply:
I'm not really sure when seek to scan change happens, 500 seemed to result in an index seek. It may need bench marking to know if its better to make 4x the queries vs the table scan, but given query at 2000 this times out frequently, I would consider it worthwhile using the smaller batch size.

Reply:
Yep will look into it, thanks!

Reply:
When testing locally and changing to 500 it unfortunately continues to use an index scan when a lot of versions are involved. A big part of the timeout issue was that it was trying to page over page sizes of 5000 for the entire content/media data, this has now been changed to be 1000 or configurable https://github.com/umbraco/Umbraco-CMS/pull/8376/files#diff-a7f4263c248ded990a4eac4c9943276fL1480  which helped with the timeouts I was getting locally. I think for individual content items or even a few items at a time these queries work ok but when we know we are going to be looking up *all* content or media I think perhaps this entire lookup process should be reconsidered. (ideally the database would just be overhauled ... but that's a whole other story)